### PR TITLE
Basic support for Unicode file names and some bugfixes

### PR DESCRIPTION
**Pull request by aafemt**
Unicode file names and pathes work at API level. Little glitches can be fixed after Pavel Zotov create test suite.
Much more must be done for unicode support in utilities command lines.

### CODE REVIEW

File: src/common/classes/fb_string.h [hunk 1]
```diff
@@ -411,10 +411,10 @@ namespace Firebird
 		}
 		void insert(size_type p0, const_pointer s, const size_type n)
 		{
```
Comment (AlexPeshkoff) [inline src/common/classes/fb_string.h]: IMO this fix should be done as separate pull request and backported to 2.5 and 3.0.

Comment (aafemt) [inline src/common/classes/fb_string.h]: No, this is regression of this branch, problem does not exist in other branches.

Comment (AlexPeshkoff) [inline src/common/classes/fb_string.h]: On 06/16/2016 07:58 PM, Dimitry Sibiryakov wrote:
>> @@ -411,10 +411,10 @@ namespace Firebird
>>   		}
>>   		void insert(size_type p0, const_pointer s, const size_type n)
>>   		{
> No, this is regression of this branch, problem does not exist in other branches.

OK, in that case no problems.

File: src/common/os/posix/os_utils.cpp [hunk 2]
```diff
@@ -393,4 +395,147 @@ void CtrlCHandler::handler(void*)
...
+			outBuf = NULL;
+		}
+	};
+
+#ifdef HAVE_ICONV_H
+	class IConverter : public Converter
+	{
+	public:
+
+		IConverter(const char* sysLocale) : ic(NULL)
```
Comment (hvlad) [inline src/common/os/posix/os_utils.cpp]: Class name "IConverter" points that class is (kind of) interface while this is not so.
Base class "Converter" is pure virtual, i.e. it is good candidate to be "IConverter" :)
Please, avoid confusion when choosing names.

Comment (aafemt) [inline src/common/os/posix/os_utils.cpp]: This is just a copy-paste from isc_file.cpp::IConv class. "I" in this case has nothing to do with interfaces.

Comment (dyemanov) [inline src/common/os/posix/os_utils.cpp]: Then it should better be renamed to IntlConverter (or whatever "I" stands for there).

Comment (aafemt) [inline src/common/os/posix/os_utils.cpp]: I leave it to one who will finish this POSIX port.

Comment (hvlad) [inline src/common/os/posix/os_utils.cpp]: > This is just a copy-paste from isc_file.cpp::IConv class.
This is not true. 
I see "class IConv" in isc_file.cpp (yes, bad name too) while you introduced 3 new classes and none of them is "IConv".
I.e. you already made changes in copied code.
If you don't want to do something - say "i don't want", no need to create empty reasoning...

> I leave it to one who will finish this POSIX port.
So, you ask us to merge half-made code and to "finish" it after you ? Cool...

Comment (aafemt) [inline src/common/os/posix/os_utils.cpp]: No. I ask for a help from someone with POSIX knowledge. Create pull request to this 
branch, I'll check and merge it. Then whole work can be merged into master branch.

Comment (hvlad) [inline src/common/os/posix/os_utils.cpp]: >  I ask for a help from someone with POSIX knowledge

It is not related with my original comment re. class names

Comment (AlexPeshkoff) [inline src/common/os/posix/os_utils.cpp]: > This is just a copy-paste from isc_file.cpp::IConv class.

class IConv is a helper for calling function iconv(): http://linux.die.net/man/3/iconv
It has nothing to do with interfaces but exactly in this form I treat it's name as good.

What about copy-paste - this is what I do not understand. Why have 2 same (may be almost same) classes with different names? You should have single class performing all actions required from iconv() and use them in all c++ modules where you need. I'm strongly against copyNpaste techology!

Comment (AlexPeshkoff) [inline src/common/os/posix/os_utils.cpp]: Re posix port. I do not see big problems with having incomplete OS-A build provided author is using OS-B. This can be fixed a bit later (specially at pre-alpha stage). Certainly it does not mean that code for OS-A may get poisoned with bad programming style.

Comment (aafemt) [inline src/common/os/posix/os_utils.cpp]: They are not same. This one is consuming 4 times less memory, but do conversion from UTF-8 
only. Different name to be avoid Linux symbol hell.

Comment (AlexPeshkoff) [inline src/common/os/posix/os_utils.cpp]: On 06/29/2016 12:59 PM, Dimitry Sibiryakov wrote:
>> +		char* convert(const char* fromBuf, size_t fromLen) override
>> +		{
>> +			return const_cast<char*>(fromBuf);
>> +		}
>> +		void destroy(char* &outBuf)
>> +		{
>> +			outBuf = NULL;
>> +		}
>> +	};
>> +
>> +#ifdef HAVE_ICONV_H
>> +	class IConverter : public Converter
>> +	{
>> +	public:
>> +
>> +		IConverter(const char* sysLocale) : ic(NULL)
> They are not same. This one is consuming 4 times less memory, but do conversion from UTF-8
> only. Different name to be avoid Linux symbol hell.

In that case one class can be derived from another.
Anyway please avoid code duplication!

**Comment by egorpugin** [comment]
Why just not use boost::filesystem and remove tons of code?
Especially when boost::filesystem was accepted into C++17.

**Comment by hvlad** [comment]
What issue the last commit (2f89d89) solved ? Just for fun ?

**Comment by aafemt** [comment]
It got rid of magic numbers before Adriano requested that.

**Comment by egorpugin** [comment]
Size of `dllExt` string is known at compile time, so it's better to use `sizeof()` operator.

**Comment by aafemt** [comment]
So, strlen also will be calculated at compile time. But in contrast with sizeof it will 
return right value, not size of pointer.

**Comment by egorpugin** [comment]
`sizeof` for `static const char dllExt[] = ".dll";`, not for `const char *`.
Why `strlen()` is compile time?

**Comment by aafemt** [comment]
Because compilers are clever nowadays.

**Comment by egorpugin** [comment]
It will probably require modern `constexpr` keyword.
By the way on line 195 of 2f89d89 you can use also `dllExt` and not `".dll"`.

File: src/auth/SecurityDatabase/LegacyManagement.epp [hunk 3]
```diff
@@ -565,7 +567,9 @@ int SecurityDatabaseManagement::execute(Firebird::CheckStatusWrapper* st, Firebi
 							if (!U.PLG$GROUP_NAME.NULL)
 								a3.printf("GroupName=%s\n", U.PLG$GROUP_NAME);
 
-							attr = a1 + a2 + a3;
+							attr = a1;
```
Comment (AlexPeshkoff) [inline src/auth/SecurityDatabase/LegacyManagement.epp]: This change has nothing to do with unicode support

**Comment by aafemt** [comment]
> This change has nothing to do with unicode support


Yes, this is result of string refactoring which you demanded.

**Comment by aafemt** [comment]
> What's a reason in using here plain char pointer instead class reference?


This routine used to be called using string literal. Creating of temporary class is 
unnecessary overhead. In rare cases when it is called with data contained in string, call 
of c_str() has no overhead.

**Comment by aafemt** [comment]
> Specially strange taken together with introduced by you assignment operator using rvalue
> reference.


This operator has no advantage on short strings which fit into internal buffer. That's why 
control of assignments is still important for performance.

File: src/common/classes/fb_string.h [hunk 4]
```diff
@@ -430,108 +314,217 @@ namespace Firebird
...
+				}
+				else // fallback to copy
+				{
+					assign(v);
+				}
+			}
+		}
 
-		AbstractString& append(const AbstractString& str)
+		void append(const AbstractString& str)
```
Comment (AlexPeshkoff) [inline src/common/classes/fb_string.h]: Rollback this massive changes from 'AbstractString&' to 'void' return type. Not related with unicode support (and violate syntax of std::basic_string).

**Comment by AlexPeshkoff** [comment]
On 10/18/16 14:00, Dimitry Sibiryakov wrote:
>> Specially strange taken together with introduced by you assignment operator using rvalue
>> reference.
>
> This operator has no advantage on short strings which fit into internal buffer. That's why
> control of assignments is still important for performance.
>
>

Anyway rollback that "optimization".

File: src/common/classes/init.cpp [hunk 5]
```diff
@@ -102,7 +102,7 @@ namespace
 			if (dladdr((void*) &allClean, &path))
 			{
 				name = path.dli_fname;
-				name += ".memdebug.log";
```
Comment (AlexPeshkoff) [inline src/common/classes/init.cpp]: Use of function call instead redefined operator does not make code better. 
Rollback such changes too.

**Comment by aafemt** [comment]
> Use of function call instead redefined operator does not make code better.


It makes code working. As was agreed in firebird-devel, the function doesn't add directory 
separator before appended string.

**Comment by AlexPeshkoff** [comment]
Dmitry, I stop review.
Please remove _yourself_ optimizations not related to unicode support.
Some of them may be discussed later in separate pull requests.

**Comment by AlexPeshkoff** [comment]
On 10/18/16 14:19, Dimitry Sibiryakov wrote:
>> Use of function call instead redefined operator does not make code better.
>
> It makes code working. As was agreed in firebird-devel, the function doesn't add directory
> separator before appended string.
>
>

OK, agreed.

File: src/common/classes/fb_string.h (hunk 4 above)
Comment (aafemt) [inline src/common/classes/fb_string.h]: These routines are private. They cannot be called from outside code and have nothing to do with public interface of std::basic_string except name.

**Comment by AlexPeshkoff** [comment]
On 10/18/16 15:31, Dimitry Sibiryakov wrote:
> aafemt commented on this pull request.
>
>
>
>>   
> -		AbstractString& append(const AbstractString& str)
> +		void append(const AbstractString& str)
>
> These routines are private.

They are protected. Am I undestand it right that
         using AbstractString::append;
in public section of derived class makes it visible from the outside world?

> They cannot be called from outside code and have nothing to do with public interface of std::basic_string except name.
>

Class string was initially designed to be able to serve as a replacement 
of basic_string<char> with extended functionality and I see no reasons 
to change this ability as long as we have no serious reasons to do it.

I.e. to be precise - not only name but name and functionality.

**Comment by aafemt** [comment]
> in public section of derived class makes it visible from the outside world?

Yes.

> I see no reasons to change this ability as long as we have no serious reasons to do it.

Isn't chained code like "a.append("abc").append("def").c_str()" prohibited by project's style guides?..

**Comment by AlexPeshkoff** [comment]
On 10/18/16 15:48, Dimitry Sibiryakov wrote:
>> in public section of derived class makes it visible from the outside world?
> Yes.
>
>> I see no reasons to change this ability as long as we have no serious reasons to do it.
> Isn't chained code like "a.append("abc").append("def").c_str()" prohibited by project's style guides?..
>

May be - but I do not know about it.
Chains of redefined operators are actively used.

**Comment by aafemt** [comment]
> Chains of redefined operators are actively used.

But Adriano called them "ugly and non-sense", didn't he?..

**Comment by AlexPeshkoff** [comment]
On 10/18/16 15:55, Dimitry Sibiryakov wrote:
>> Chains of redefined operators are actively used.
> But Adriano called them "ugly and non-sense", didn't he?..
>

Do not think so - he is actively them. For example, in SQL builder.
But that all goes offtopic here.

File: src/burp/restore.epp [hunk 6]
```diff
@@ -657,7 +658,7 @@ void add_files(BurpGlobals* tdgbl, const char* file_name)
 
 	for (burp_fil* file = tdgbl->gbl_sw_files; file; file = file->fil_next)
 	{
-		if (file->fil_name != file_name)
```
Comment (AlexPeshkoff) [inline src/burp/restore.epp]: Why this change?
Are we missing operator!=()?

**Comment by aafemt** [comment]
> Are we missing operator!=()?

  Yes.
